pkg/reconciler/v1alpha1/broker/broker.go

```diff
@@ -0,0 +1,442 @@
+
+	if err := r.client.Get(ctx, objectKey, latestBroker); err != nil {
+		return nil, err
+	}
+
+	brokerChanged := false
+
+	if !equality.Semantic.DeepEqual(latestBroker.Finalizers, broker.Finalizers) {
+		latestBroker.SetFinalizers(broker.ObjectMeta.Finalizers)
+		if err := r.client.Update(ctx, latestBroker); err != nil {
+			return nil, err
+		}
+		brokerChanged = true
+	}
+
+	if equality.Semantic.DeepEqual(latestBroker.Status, broker.Status) {
+		return latestBroker, nil
+	}
+
+	if brokerChanged {
+		// Re-fetch.
+		latestBroker = &v1alpha1.Broker{}
+		if err := r.client.Get(ctx, objectKey, latestBroker); err != nil {
+			return nil, err
+		}
+	}
+
+	latestBroker.Status = broker.Status
+	if err := r.client.Status().Update(ctx, latestBroker); err != nil {
+		return nil, err
+	}
+
+	return latestBroker, nil
+}
+
+// reconcileFilterDeployment reconciles Broker's 'b' filter deployment.
+func (r *reconciler) reconcileFilterDeployment(ctx context.Context, b *v1alpha1.Broker) (*v1.Deployment, error) {
+	expected := resources.MakeFilterDeployment(&resources.FilterArgs{
+		Broker:             b,
+		Image:              r.filterImage,
+		ServiceAccountName: r.filterServiceAccountName,
+	})
+	return r.reconcileDeployment(ctx, expected)
+}
+
+// reconcileFilterService reconciles Broker's 'b' filter service.
+func (r *reconciler) reconcileFilterService(ctx context.Context, b *v1alpha1.Broker) (*corev1.Service, error) {
+	expected := resources.MakeFilterService(b)
+	return r.reconcileService(ctx, expected)
+}
+
+// reconcileChannel reconciles Broker's 'b' underlying channel.
+func (r *reconciler) reconcileChannel(ctx context.Context, b *v1alpha1.Broker) (*v1alpha1.Channel, error) {
+	c, err := r.getChannel(ctx, b)
+	// If the resource doesn't exist, we'll create it
+	if k8serrors.IsNotFound(err) {
+		c = newChannel(b)
+		err = r.client.Create(ctx, c)
+		if err != nil {
+			return nil, err
+		}
+		return c, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	// TODO Determine if we want to update spec (maybe just args?).
+	// Update Channel if it has changed. Note that we need to both ignore the real Channel's
+	// subscribable section and if we need to update the real Channel, retain it.
+	//expected.Spec.Subscribable = c.Spec.Subscribable
+	//if !equality.Semantic.DeepDerivative(expected.Spec, c.Spec) {
+	//	c.Spec = expected.Spec
+	//	err = r.client.Update(ctx, c)
+	//	if err != nil {
+	//		return nil, err
+	//	}
+	//}
+	return c, nil
+}
+
+// getChannel returns the Channel object for Broker 'b' if exists, otherwise it returns an error.
+func (r *reconciler) getChannel(ctx context.Context, b *v1alpha1.Broker) (*v1alpha1.Channel, error) {
+	list := &v1alpha1.ChannelList{}
+	opts := &runtimeclient.ListOptions{
+		Namespace:     b.Namespace,
+		LabelSelector: labels.SelectorFromSet(ChannelLabels(b)),
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
```

Comment:
I don't think this is needed with controller-runtime List
Q: Is it possible to use a FieldSelector here to filter by ownerReference?

Reply:
To the best of my knowledge, no we can't use a `FieldSelector` on anything other than `name` and `namespace`. See https://github.com/kubernetes/kubernetes/issues/53459#issuecomment-434040642.

Reply:
What do you mean by "I don't think this is need with controller-runtime List"?

Does it mean that controller-runtime List will never need to page, it will always return everything at once, regardless of the number that match?

Now, in this case we don't actually look at the continue token, so it clearly isn't needed here, but I still feel safer leaving it in (in case someone adds the page logic later).